File: 138.java
class Solution {
    public Node copyRandomList(Node head) {
        Node cursor = head;
        while (cursor != null) {
            Node newNode = new Node(cursor.val);
            newNode.next = cursor.next;
            cursor.next = newNode;
            cursor = newNode.next;
        }

        cursor = head;
        while (cursor != null) {
            if (cursor.random != null) {
                cursor.next.random = cursor.random.next;
            }
            cursor = cursor.next.next;
        }

        Node dummyHead = new Node(0), dummyCursor = dummyHead;
        cursor = head;
        while (cursor != null) {
            dummyCursor.next = cursor.next;
            dummyCursor = dummyCursor.next;

            cursor.next = cursor.next.next;
            cursor = cursor.next;
        }
        return dummyHead.next;
    }
}

File: Python/138.py
from typing import Optional


class Node:
    def __init__(self, x: int, next: "Node" = None, random: "Node" = None):
        self.val = int(x)
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: "Optional[Node]") -> "Optional[Node]":
        cursor = head
        while cursor:
            new_node = Node(cursor.val, cursor.next, cursor.random)
            cursor.next = new_node
            cursor = new_node.next

        cursor = head
        while cursor:
            if cursor.next.random != None:
                cursor.next.random = cursor.next.random.next
            cursor = cursor.next.next

        dummy = dummy_cursor = Node(-1)
        cursor = head
        while cursor:
            dummy_cursor.next = cursor.next
            cursor.next = cursor.next.next
            cursor = cursor.next
            dummy_cursor = dummy_cursor.next

        return dummy.next
